File: system_monitor/collector/gpio.py
from __future__ import annotations

from .base import Sensor, SensorReading


class Dht22Sensor(Sensor):
    def read(self) -> SensorReading:
        pin = int(self.config.params.get("pin", 4))
        try:
            from gpiozero import DHT22  # type: ignore[import-untyped]
            from gpiozero.pins.native import NativeFactory  # type: ignore[import-untyped]
        except ImportError:
            return SensorReading(
                sensor_id=self.id,
                value=None,
                status="error",
                error="Установите gpiozero: pip install -r requirements-pi.txt",
            )

        try:
            NativeFactory()
            sensor = DHT22(pin)
            value = sensor.temperature
        except Exception as exc:
            return SensorReading(
                sensor_id=self.id,
                value=None,
                status="error",
                error=str(exc),
            )

        if value is None:
            return SensorReading(
                sensor_id=self.id,
                value=None,
                status="unknown",
                error="Не удалось считать DHT22",
            )

        value = round(float(value), 1)
        return SensorReading(
            sensor_id=self.id,
            value=value,
            status=self.evaluate_status(value),
        )

File: internal/collector/gpio.go
//go:build linux && arm

package collector

import (
	"fmt"
	"math"
)

type dht22Sensor struct{ baseSensor }

func (s dht22Sensor) Read() SensorReading {
	pin := 4
	if s.cfg.Params != nil {
		if p, ok := s.cfg.Params["pin"].(int); ok {
			pin = p
		} else if p, ok := s.cfg.Params["pin"].(float64); ok {
			pin = int(p)
		}
	}
	value, err := readDHT22(pin)
	if err != nil {
		return SensorReading{SensorID: s.ID(), Status: "error", Error: err.Error()}
	}
	v := math.Round(value*10) / 10
	return SensorReading{SensorID: s.ID(), Value: &v, Status: evaluateStatus(s.cfg, &v)}
}

func readDHT22(pin int) (float64, error) {
	return 0, fmt.Errorf("DHT22: установите periph.io на Raspberry Pi")
}
